File: app/src/main/java/com/afollestad/photoaffix/views/SquareFrameLayout.java
package com.afollestad.photoaffix.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import com.afollestad.photoaffix.R;

/** @author Aidan Follestad (afollestad) */
public class SquareFrameLayout extends FrameLayout {

  private Paint edgePaint;
  private int borderRadius;

  public SquareFrameLayout(Context context) {
    super(context);
    init();
  }

  public SquareFrameLayout(Context context, AttributeSet attrs) {
    super(context, attrs);
    init();
  }

  public SquareFrameLayout(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init();
  }

  private void init() {
    borderRadius = (int) getResources().getDimension(R.dimen.default_square_border_radius);
    edgePaint = new Paint();
    edgePaint.setAntiAlias(true);
    edgePaint.setStyle(Paint.Style.STROKE);
    edgePaint.setColor(ContextCompat.getColor(getContext(), R.color.browseButtonBorder));
    edgePaint.setStrokeWidth(borderRadius);
  }

  @Override
  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
    //noinspection SuspiciousNameCombination
    super.onMeasure(widthMeasureSpec, widthMeasureSpec);
  }

  @Override
  protected void onDraw(Canvas canvas) {
    super.onDraw(canvas);
    int left = borderRadius;
    int top = borderRadius;
    int bottom = getMeasuredHeight() - borderRadius;
    int right = getMeasuredWidth() - borderRadius;
    canvas.drawRect(left, top, right, bottom, edgePaint);
  }
}

File: app/src/main/java/com/afollestad/photoaffix/components/SquareFrameLayout.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.photoaffix.components

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Paint.Style.STROKE
import android.util.AttributeSet
import android.widget.FrameLayout
import androidx.core.content.ContextCompat.getColor
import com.afollestad.photoaffix.R

/** @author Aidan Follestad (afollestad)
 */
class SquareFrameLayout(
  context: Context,
  attrs: AttributeSet? = null
) : FrameLayout(context, attrs) {

  private val edgePaint: Paint
  private val borderRadius: Int =
    context.resources.getDimension(R.dimen.default_square_border_radius)
        .toInt()

  init {
    edgePaint = Paint().apply {
      isAntiAlias = true
      style = STROKE
      color = getColor(context, R.color.browseButtonBorder)
      strokeWidth = borderRadius.toFloat()
    }
  }

  override fun onMeasure(
    widthMeasureSpec: Int,
    heightMeasureSpec: Int
  ) {
    super.onMeasure(widthMeasureSpec, widthMeasureSpec)
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    val left = borderRadius
    val top = borderRadius
    val bottom = measuredHeight - borderRadius
    val right = measuredWidth - borderRadius
    canvas.drawRect(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat(), edgePaint)
  }
}
